core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -777,6 +787,75 @@ NavigableMap<Long, InFlightBatch> cachedState() {
         return new ConcurrentSkipListMap<>(cachedState);
     }
 
+    // TODO: maxDeliveryCount should be utilized here once it is implemented
+    /**
+     * Apply acquisition lock to acquired records.
+     * @param memberId The member id of the client that is putting the acquisition lock.
+     * @param baseOffset The base offset of the acquired records.
+     * @param lastOffset The last offset of the acquired records.
+     * @param cachedStateEntry The cached state map entry for the acquired records which contains the entire batch
+     *                         from base offset to last offset.
+     */
+    private void startAcquisitionLockTimer(String memberId, long baseOffset, long lastOffset, Map.Entry<Long, InFlightBatch> cachedStateEntry) {
+        timer.add(new TimerTask(recordLockDurationMs) {
+            // Runs when the acquisition lock timer expires. We would then be releasing the acquired records.
+            @Override
+            public void run() {
+                lock.writeLock().lock();
+                try {
+                    InFlightBatch inFlightBatch = cachedStateEntry.getValue();
+                    long localNextFetchOffset = nextFetchOffset;
+                    // Case when the state of complete batch is valid
+                    if (inFlightBatch.offsetState == null) {
```

Comment:
Hi @apoorvmittal10 , the situation you talked about is covered in this test [here](https://github.com/confluentinc/kafka/pull/1157/files#diff-1e8cade5bbb07e7982c7c7a48cd5f49f9cda5324b73fbbe0ef1332c71a645387R1541)

Reply:
>It's not very straightforward for me to get the implementation when I do not find how you will cancel or amend the task on successful batch/subset acknowledgement.

As discussed, this code is correct but inefficient because it does not give the ability to cancel tasks. You can only add tasks and cases where partial/complete acknowledgement comes are handled within the `run()` function [here](https://github.com/confluentinc/kafka/pull/1157/files#diff-4ef2490e40ca0c63d51bf46caf866690472805e34d3752a919217b28f97045bcR803).